runelite-client/src/main/java/net/runelite/client/plugins/inferno/InfernoOverlay.java

```diff
@@ -1,112 +1,480 @@
+	}
+
+	private void renderIndividualTilesSafespots(Graphics2D graphics)
+	{
+		for (WorldPoint worldPoint : plugin.getSafeSpotMap().keySet())
+		{
+			final int safeSpotId = plugin.getSafeSpotMap().get(worldPoint);
+
+			if (safeSpotId > 6)
+			{
+				continue;
+			}
+
+			final LocalPoint localPoint = LocalPoint.fromWorld(client, worldPoint);
+
+			if (localPoint == null)
+			{
+				continue;
+			}
+
+			final Polygon tilePoly = Perspective.getCanvasTilePoly(client, localPoint);
+
+			if (tilePoly == null)
+			{
+				continue;
+			}
+
+			Color color;
+			switch (safeSpotId)
+			{
+				case 0:
+					color = Color.WHITE;
+					break;
+				case 1:
+					color = Color.RED;
+					break;
+				case 2:
+					color = Color.GREEN;
+					break;
+				case 3:
+					color = Color.BLUE;
+					break;
+				case 4:
+					color = new Color(255, 255, 0);
+					break;
+				case 5:
+					color = new Color(255, 0, 255);
+					break;
+				case 6:
+					color = new Color(0, 255, 255);
+					break;
+				default:
+					continue;
+			}
+
+			OverlayUtil.renderPolygon(graphics, tilePoly, color);
+		}
+	}
+
+	private void renderTicksOnNpc(Graphics2D graphics, InfernoNPC infernoNPC)
+	{
+		final Color color = (infernoNPC.getTicksTillNextAttack() == 1
+			|| (infernoNPC.getType() == InfernoNPC.Type.BLOB && infernoNPC.getTicksTillNextAttack() == 4))
+			? infernoNPC.getNextAttack().getCriticalColor() : infernoNPC.getNextAttack().getNormalColor();
+		final Point canvasPoint = infernoNPC.getNpc().getCanvasTextLocation(
+			graphics, String.valueOf(infernoNPC.getTicksTillNextAttack()), 0);
+		OverlayUtil.renderTextLocation(graphics, String.valueOf(infernoNPC.getTicksTillNextAttack()),
+			plugin.getTextSize(), plugin.getFontStyle().getFont(), color, canvasPoint, false, 0);
+	}
+
+	private void renderNpcLocation(Graphics2D graphics, InfernoNPC infernoNPC)
+	{
+		final LocalPoint localPoint = LocalPoint.fromWorld(client, infernoNPC.getNpc().getWorldLocation());
+
+		if (localPoint != null)
+		{
+			final Polygon tilePolygon = Perspective.getCanvasTilePoly(client, localPoint);
+
+			if (tilePolygon != null)
+			{
+				OverlayUtil.renderPolygon(graphics, tilePolygon, Color.BLUE);
+			}
 		}
-		return null;
 	}
 
-	// renders text location
-	private static void renderTextLocation(Graphics2D graphics, InfernoNPC actor, String text, Color color)
+	private void renderDescendingBoxes(Graphics2D graphics)
 	{
-		graphics.setFont(new Font("Arial", Font.BOLD, 15));
-		Point textLocation = actor.getNpc().getCanvasTextLocation(graphics, text, actor.textLocHeight + 40);
-		if (Strings.isNullOrEmpty(text))
+		for (Integer tick : plugin.getUpcomingAttacks().keySet())
 		{
-			return;
+			final HashMap<InfernoNPC.Attack, Integer> attackPriority = plugin.getUpcomingAttacks().get(tick);
+			int bestPriority = 999;
+			InfernoNPC.Attack bestAttack = null;
+
+			for (InfernoNPC.Attack currentAttack : attackPriority.keySet())
+			{
+				final int currentPriority = attackPriority.get(currentAttack);
```

Comment:
this should just iterate over the entryset, rather than iterating over the keyset and getting the value